with_plan_module() raises TypeError when experiment plan signature has a union type expression

On upgrading hyperion to python 3.11, we encountered an issue running the tests, where our context is initialised, our experiment plans have union type expressions in the signature e.g.:

```
def flyscan_xray_centre(
    composite: FlyScanXRayCentreComposite,
    parameters: ThreeDGridScan | GridscanInternalParameters | Any,
) -> MsgGenerator:
```
This caused our CI to fail with the following `TypeError: type 'types.UnionType' is not subscriptable` exception:

https://github.com/DiamondLightSource/hyperion/actions/runs/8800638512/job/24152242508

```
src/hyperion/utils/context.py:77: in setup_context
    context.with_plan_module(hyperion_plans)
/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/blueapi/core/context.py:136: in with_plan_module
    self.plan(obj)
/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/blueapi/core/context.py:167: in plan
    **self._type_spec_for_function(plan),
/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/blueapi/core/context.py:269: in _type_spec_for_function
    self._convert_type(arg_type),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = BlueskyContext(run_engine=<bluesky.run_engine.RunEngine object at 0x7f340a219050>, plan_wrappers=[], plans={}, devices={}, plan_functions={}, sim=False, _reference_cache={})
typ = hyperion.parameters.gridscan.ThreeDGridScan | hyperion.parameters.plan_specific.gridscan_internal_params.GridscanInternalParameters | typing.Any

    def _convert_type(self, typ: Type) -> Type:
        """
        Recursively convert a type to something that can be deserialised by
        pydantic. Bluesky protocols (and types that extend them) are replaced
        with an intermediate reference types that allows the current context to
        be used to look up an existing device when deserialising device ID
        strings.
    
        Other types are returned as passed in.
    
        Args:
            typ: The type that is required - potentially referencing Bluesky protocols
    
        Returns:
            A Type that can be deserialised by Pydantic
        """
        if typ in BLUESKY_PROTOCOLS or any(
            isinstance(typ, dev) for dev in BLUESKY_PROTOCOLS
        ):
            return self._reference(typ)
        args = get_args(typ)
        if args:
            new_types = tuple(self._convert_type(i) for i in args)
            root = get_origin(typ)
>           return root[new_types] if root else typ
E           TypeError: type 'types.UnionType' is not subscriptable

/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/blueapi/core/context.py:298: TypeError
```

We can work around this for now by changing our signatures to Union[X,Y] instead of X|Y

--- a/src/blueapi/core/context.py
+++ b/src/blueapi/core/context.py
@@ -4,15 +4,8 @@ from collections.abc import Callable, Sequence
 from dataclasses import dataclass, field
 from importlib import import_module
 from inspect import Parameter, signature
-from types import ModuleType
-from typing import (
-    Any,
-    Generic,
-    TypeVar,
-    get_args,
-    get_origin,
-    get_type_hints,
-)
+from types import ModuleType, UnionType
+from typing import Any, Generic, TypeVar, Union, get_args, get_origin, get_type_hints
 
 from bluesky.run_engine import RunEngine, call_in_bluesky_event_loop
 from pydantic import create_model
@@ -264,7 +257,7 @@ class BlueskyContext:
             )
         return new_args
 
-    def _convert_type(self, typ: type) -> type:
+    def _convert_type(self, typ: type | Any) -> type:
         """
         Recursively convert a type to something that can be deserialised by
         pydantic. Bluesky protocols (and types that extend them) are replaced
@@ -288,6 +281,8 @@ class BlueskyContext:
         if args:
             new_types = tuple(self._convert_type(i) for i in args)
             root = get_origin(typ)
+            if root == UnionType:
+                root = Union
             return root[new_types] if root else typ
         return typ
 

--- a/tests/core/test_context.py
+++ b/tests/core/test_context.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+from typing import Union
 from unittest.mock import patch
 
 import pytest
@@ -271,6 +272,24 @@ def test_reference_type_conversion(empty_context: BlueskyContext) -> None:
     assert (
         empty_context._convert_type(dict[Movable, list[tuple[int, Movable]]])
         == dict[movable_ref, list[tuple[int, movable_ref]]]  # type: ignore
+    )
+
+
+def test_reference_type_conversion_union(empty_context: BlueskyContext) -> None:
+    movable_ref: type = empty_context._reference(Movable)
+    assert empty_context._convert_type(Movable) == movable_ref
+    assert (
+        empty_context._convert_type(Union[Movable, int]) == Union[movable_ref, int]  # noqa # type: ignore
+    )
+
+
+def test_reference_type_conversion_new_style_union(
+    empty_context: BlueskyContext,
+) -> None:
+    movable_ref: type = empty_context._reference(Movable)
+    assert empty_context._convert_type(Movable) == movable_ref
+    assert (
+        empty_context._convert_type(Movable | int) == movable_ref | int  # type: ignore
     )
 
 

